src/csharp/Microsoft.Spark.Worker/Utils/AssemblyLoaderHelper.cs

```diff
@@ -0,0 +1,163 @@
+        /// </summary>
+        /// <param name="stageId">The current Stage ID</param>
+        internal static void RegisterAssemblyHandler(int stageId)
+        {
+            if (!EnvironmentUtils.GetEnvironmentVariableAsBool("DOTNET_SPARK_REPL_MODE") ||
+                (stageId == s_stageId))
+            {
+                return;
+            }
+
+            // For a given stage, it is sufficient to instantiate one DependencyProvider.
+            // However, the Worker process may be reused between stages. New nuget dependencies
+            // may be introduced between stages and a new DependencyProvider will need to be
+            // created that can resolve them.
+            lock (s_lock)
+            {
+                if (stageId == s_stageId)
+                {
+                    return;
+                }
+                s_stageId = stageId;
+
+                string sparkFilesPath = SparkFiles.GetRootDirectory();
+                string metadataFile =
+                    FindHighestFile(sparkFilesPath, DependencyProviderUtils.FilePattern);
+
+                if (string.IsNullOrEmpty(metadataFile) || metadataFile.Equals(s_lastFileRead))
+                {
+                    return;
+                }
+                s_lastFileRead = metadataFile;
+
+                using FileStream fileStream = File.OpenRead(metadataFile);
+                BinaryFormatter formatter = new BinaryFormatter();
+                DependencyProviderUtils.Metadata metadata =
+                    (DependencyProviderUtils.Metadata)formatter.Deserialize(fileStream);
+
+                string unpackPath =
+                    Path.Combine(Directory.GetCurrentDirectory(), Path.Combine(".nuget", "packages"));
+                Directory.CreateDirectory(unpackPath);
+
+                IEnumerable<string> AssemblyProbingPaths()
+                {
+                    foreach (string dependency in metadata.AssemblyProbingPaths)
+                    {
+                        yield return Path.Combine(unpackPath, dependency);
+                    }
+                }
+
+                IEnumerable<string> NativeProbingRoots()
+                {
+                    foreach (string dependency in metadata.NativeProbingPaths)
+                    {
+                        yield return Path.Combine(unpackPath, dependency);
+                    }
+                }
+
+                UnpackPackages(sparkFilesPath, unpackPath, metadata.NuGets);
+
+                (s_dependencyProvider as IDisposable)?.Dispose();
+                s_dependencyProvider = new DependencyProvider(AssemblyProbingPaths, NativeProbingRoots);
+            }
+        }
+
+        private static string FindHighestFile(string src, string pattern)
+        {
+            string[] files = Directory.GetFiles(src, pattern);
+            if (files.Length > 0)
+            {
+                Array.Sort(files);
+                return files.Last();
```

Comment:
The highest file should have accumulated info right? In that case should we try to remove the old files proactively or no?

Reply:
Correct.  The highest file should contain the super set of all nuget, assembly, and native probing paths.

We can remove them proactively, but I don't think we'll have that many files.  There would only be a new file for every cell that introduces a new nuget package.